src/ln/channelmanager.rs

```diff
@@ -1479,7 +1518,81 @@ impl ChannelManager {
 	/// ChannelMonitorUpdateErr::TemporaryFailure was returned from a channel monitor update
 	/// operation.
 	pub fn test_restore_channel_monitor(&self) {
-		unimplemented!();
+		let mut new_events = Vec::new();
+		let mut close_results = Vec::new();
+		let mut htlc_forwards = Vec::new();
+		let mut htlc_failures = Vec::new();
+
+		let mut channel_lock = self.channel_state.lock().unwrap();
+		let channel_state = channel_lock.borrow_parts();
+		let short_to_id = channel_state.short_to_id;
+		channel_state.by_id.retain(|_, channel| {
+			if channel.is_awaiting_monitor_update() {
+				let chan_monitor = channel.channel_monitor();
+				if let Err(e) = self.monitor.add_update_monitor(chan_monitor.get_funding_txo().unwrap(), chan_monitor) {
+					match e {
+						ChannelMonitorUpdateErr::PermanentFailure => {
+							if let Some(short_id) = channel.get_short_channel_id() {
+								short_to_id.remove(&short_id);
+							}
+							close_results.push(channel.force_shutdown());
+							if let Ok(update) = self.get_channel_update(&channel) {
+								new_events.push(events::Event::BroadcastChannelUpdate {
+									msg: update
+								});
+							}
+							false
+						},
+						ChannelMonitorUpdateErr::TemporaryFailure => true,
+					}
+				} else {
+					let (raa, commitment_update, order, pending_forwards, mut pending_failures) = channel.monitor_updating_restored();
+					if !pending_forwards.is_empty() {
+						htlc_forwards.push((channel.get_short_channel_id().expect("We can't have pending forwards before funding confirmation"), pending_forwards));
+					}
+					htlc_failures.append(&mut pending_failures);
+
+					macro_rules! handle_cs { () => {
+						if let Some(update) = commitment_update {
+							new_events.push(events::Event::UpdateHTLCs {
+								node_id: channel.get_their_node_id(),
+								updates: update,
+							});
+						}
+					} }
+					macro_rules! handle_raa { () => {
+						if let Some(revoke_and_ack) = raa {
+							new_events.push(events::Event::SendRevokeAndACK {
+								node_id: channel.get_their_node_id(),
+								msg: revoke_and_ack,
+							});
+						}
+					} }
+					match order {
+						RAACommitmentOrder::CommitmentFirst => {
+							handle_cs!();
+							handle_raa!();
+						},
+						RAACommitmentOrder::RevokeAndACKFirst => {
+							handle_raa!();
+							handle_cs!();
+						},
+					}
+					true
+				}
+			} else { true }
+		});
+
+		for failure in htlc_failures.drain(..) {
+			self.fail_htlc_backwards_internal(self.channel_state.lock().unwrap(), failure.0, &failure.1, failure.2);
```

Comment:
there is no deadlock with the one at L1526 here ? hmm when this last one is dropped?
Same as upper comment on finish_force_close, maybe we should modify some fn parameters in the future to pass down lock, depends how per-channel locks will look like.

Reply:
Oops, good catch! Luckily not triggerable since the only handled case is on send_payment, and any future handled catches should have tested this.